pkg/pilot/genericpilot/controller_test.go

```diff
@@ -0,0 +1,268 @@
+func newDefaultProcessAdapter() process.Interface {
+	return &process.FakeAdapter{
+		RunningFunc: func() bool { return true },
+		StringFunc:  func() string { return "ps-adapter" },
+	}
+}
+
+func TestSync(t *testing.T) {
+	tests := []*fixture{
+		{
+			name:                   "successful run of Sync on 'this' pilot should return no error and update status",
+			key:                    "namespace/name",
+			err:                    false,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("name", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("name", "namespace")},
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name:                   "should not Start a subprocess when syncing a non-this pilot",
+			key:                    "namespace/not-this-pilot",
+			err:                    false,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/not-this-pilot"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("not-this-pilot", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("not-this-pilot", "namespace")},
+		},
+		{
+			name:                   "failed genericpilot consumer sync function should fail sync",
+			key:                    "namespace/name",
+			err:                    true,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("name", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("name", "namespace")},
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name:               "missing cachedThisPilot and empty clientset should fail Sync()",
+			key:                "namespace/name",
+			err:                true,
+			thisPilotName:      "name",
+			thisPilotNamespace: "namespace",
+		},
+		{
+			name: "should reuse cached 'this' pilot resource, and should fail to update status",
+			key:  "namespace/name",
+			err:  true,
+			syncFuncExpectedPilot:  newDummyPilot("name", "namespace"),
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
```

Comment:
The test title says `should fail to update status` but the expected `navActions` include a status update.
Perhaps the test title could be reworded.

Reply:
👍 basically, if the resource doesn't exist, it *will* attempt to update the status (hence the expected action), but will fail with an error (hence `err: true`). I could update this to say it "should attempt to update status and fail"